### ISSUE TITLE
Systemd "After" rules for correct node starting

### ISSUE THREAD

**Issue by chief93** [kubernetes/kubeadm#2159]
## FEATURE REQUEST

Hi everyone! Following by [this issue](https://github.com/kubernetes/kubernetes/issues/90594), where I've found a potential root cause of start-up misconfiguration, I propose to include the following instructions to the beginning of the `kubeadm` `systemd` service in the official `kubeadm` packages like `.deb`, `.rpm` and other:

```ini
[Unit]
Description=Kubernetes
After=syslog.target
After=network.target
```

In mentioned issue I encountered many weird errors when without proposed solution, the `kubelet` could not recognize the primary (physical) network interface. Like I mentioned there, I just tested this solution for my cluster and all working as expected.

## Versions

**kubeadm version**: `1.18.1`
**Environment**:
- **Kubernetes version**: `1.18.1`
- **Cloud provider or hardware configuration**: `LXC` containers for Kubernetes nodes, runned on physical bare-metal server
- **OS**: Ubuntu `18.04`
- **Kernel**: `4.15.0-96-generic #97-Ubuntu SMP Wed Apr 1 03:25:46 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`
- **Others**: `CNI` plug-in is `kube-router`, bu seems like this solution will be also helpful for `Flannel`

**Comment by neolit123** [comment]
hello,

> In mentioned issue I encountered many weird errors when without proposed solution, the kubelet could not recognize the primary (physical) network interface. Like I mentioned there, I just tested this solution for my cluster and all working as expected.

when the kubelet DEB/RPM service is deployed on the user host it starts crashlooping because it's missing a --config file. kubeadm join/init writes the file and restarts the kubelet service, at which point the kubelet should be able to pick the right interface too.

so from my above comment, i'm confused why waiting for syslog.target and network.target is required, given if the initial kubelet service service runs are always expected to fail, until the user runs kubeadm init/join.

maybe the host network is not starting at all, in your setup and you are using the kubelet service to kick-start it? perhaps i'm missing something else?

> Others: CNI plug-in is kube-router, bu seems like this solution will be also helpful for Flannel

i've seen quite a big number of issues related to flannel and kube-router lately....usually calico and weavenet don't have weird errors usually.

also i agree with this comment:
https://github.com/kubernetes/kubernetes/issues/90594#issuecomment-635757628

from my POV, LXC is pretty much unsupported and i don't think we have e2e tests for it.

/priority awaiting-more-evidence
cc @rosti you had a recent proposal for systemd file updates.

**Comment by chief93** [comment]
@neolit123 first of all - thanks for reply!

About your questions - yes, you miss the idea that the LXC itself IS NOT THE CAUSE OF THIS PROBLEM EITHER! The main cause is, as I mentioned it here and in attached issue, is the ORDER OF STARTING of the `kubelet` `systemd` service! The proposed solution is quite simple and is absolutely required for any software that uses the networking, obviously.

As I mentioned in other issues, I can't use `Calico` because of overlapping of my home network address pool. And after using of very stable `kube-router` I even don't want to change it to something else.

Again about the `systemd` - in it's docs is pretty clear said that any apps, that require the networking is need to be started AFTER the networking service, it's a enormously obvious statement!

Summarizing (if anybody didn't read the attached issue at all):
1. Hypervisor OS setup normally
2. Worker nodes setup normally
3. Recommended Kubernetes installation procedure executed
4. Encounter of spamming of ARP tables by `Flannel` and even not starting it from scratch at all, only after non-trivial config patches
5. Absolutely normally start up from scratch of the `kube-router`
6. The wrong depending of the primary network interface (**PROPOSED IN THIS ISSUE SOLUTION FIXES THIS FOR 146%**)
7. BUG in the `kube-proxy`, then full elimination of `kube-proxy` without any negative impact of the cluster functionality (I'm wondering, for what purposes ever exists such bugged software, only for spamming of its errors to the logs?!) described in [this issue](https://github.com/kubernetes/kubernetes/issues/90083)

**Comment by chief93** [comment]
If anybody can't understand how such can happen, let me describe this process:
1. `lxc-start node1`
---inside node1---
2. emulation of the OS boot
3. previous step require the starting of the `systemd` and all it's inheritors
4. [my case] node1 send DHCP request
5. [my case] waiting for the DHCP response
6. [my case] finishing starting up of `kubelet` **(UNEXPECTED BEHAVIOR: primary network interface does not obtained the address still)**
7. [my case] DHCP response
8. bugged behabior of the cluster

Proposed solution fix the order of the starting: `kubelet` wait for the `eth0` to obtain the address from DHCP server, then starts. It is ABSOLUTELY NORMAL AND EXPECTED TO BE IN ANY CASE EITHER, even if the `eth0` is configured manually, because in manual case the interface is ready immediately, that is the only, very obvious, difference.

**Comment by neolit123** [comment]
at "kubeadm init/join" invocation time this does not make sense as explained above, unless your `eth0` is not configured yet at that point, which is bad and can cause issues with kubeadm control-plane nodes, as they also probe the primary interface outside of the kubelet logic.

now, if a node configured by kubeadm is restarted, it is true that the kubelet needs an IP from the primary interface before it starts, but i'd expected this IP to be the same as it was before the restart.
to pin the IP, the kubelet supports a --node-ip flag.

also here:
https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/
it says:

> network.target has very little meaning during start-up. It only indicates that the network management stack is up after it has been reached. Whether any network interfaces are already configured when it is reached is undefined. Its primary purpose is for ordering things properly at shutdown: since the shutdown ordering of units in systemd is the reverse of the startup ordering, any unit that is order After=network.target can be sure that it is stopped before the network is shut down if the system is powered off.

note the "Whether any network interfaces are already configured when it is reached is undefined" part. what are your comments on the above?

if i'm reading this correctly, by utilizing "After=network.target", there is no guarantee that "eth0" would be configured before the kubelet service starts.

----

separate question, why add "After=syslog.target"?

reading online it was removed in later versions of systemd:
https://opensuse-packaging.opensuse.narkive.com/MepYnewS/reminder-for-systemd-services-do-not-use-syslog-target-anymore

> With recent (>202) systemd release, syslog.target is no longer provided:

**Comment by chief93** [comment]
@neolit123 It's a nonsense to rely on the "IP hat cwill not change". In real world you can not rely on this in any situation!

About your notes:
1. It seems like you don't fully read of the cite, that you provided: there is also the `network-online.target`, that do exactly what you "doubt in" for a regular `network.target`. I said, that in my case the regular `network.target` worked perfectly. More of that - for my `systemd` apps, that require networking, I always set this flag also, and always all works as expected, non mateer of where and how the app is deployed.
2. About `syslog.target` - this is only ensurance that all `kubelet` logs will be reach to the `journalctl`, as expected. For puposes of this issue this doesn't matter, but I'm shocked that nobody either thought about such obvious actions to be done.

**Comment by rosti** [comment]
I think that you are both talking for different scenarios here. One thing needs to be made clear here. The kubeadm maintainers don't maintain the kubelet packaging and systemd service file (although we can and do send PRs from time to time to update them).

The case which @neolit123 is mentioning is the `kubeadm init/join/upgrade` one. kubeadm itself is not designed to be invoked during system startup.

The case which @chief93 covers is that of normal kubelet startup at boot (after kubeadm is called).

Now on the topic, I recently sent a PR to have the systemd targets updated. Now they point to `network-online.target`, but this would likely be released in 1.19.
The `syslog.target` is deprecated in systemd v35 and newer. Unit files should not depend on it and some Linux distros (I thing OpenSUSE for example, @ereslibre can correct me) don't even provide a `syslog.target` anymore.

About dynamic IPs. Kubernetes in general is not really tuned to support changing IPs. Certificates, config files, command lines, etc. all have lists of IPs of peers, etc. embedded in them.
Although there are ways to avoid doing these (using DNS tricks) I am not aware of any Kubernetes deployment tool that supports changing IPs. Even if you use DHCP to obtain your IPs (which most people do) they use static leases as well so that IPs won't change. This seems to be the most popular solution among production clusters.

**Comment by neolit123** [comment]
@chief93 

> there is also the network-online.target

i did read about `network-online.target`, but was asking why you are proposing `network.target` instead.

@rosti

> Now on the topic, I recently sent a PR to have the systemd targets updated. Now they point to network-online.target, but this would likely be released in 1.19.

i forgot this change is there already:
https://github.com/kubernetes/release/blob/master/cmd/kubepkg/templates/latest/rpm/kubelet/kubelet.service#L5

with `syslog.target` deprecated i don't think there are remaining actions for this ticket.
the `network-online.target` change should not be backported to <1.19.

thanks
/close

**Comment by chief93** [comment]
@rosti thank you very much for reply! As you correctly mentioned, I am talking about the `kubelet` startup procedure. 

About the DCHP static leases - that's exactly my setup - Kubenetes nodes, and even LXC hypervsior, obtain their static assigned addresses by DHCP leasings.

And thank you about mentioning of the `syslog.target` deprecation - I just used this for few years in my apps, so even didn't know that it is already deprecated.

About mentioned `1.19` version - it's ok, because the main goal is to these patches ever be set :)
